Compiler/src/tfuncs.jl

```diff
@@ -2244,7 +2250,16 @@ function _builtin_nothrow(𝕃::AbstractLattice, @nospecialize(f::Builtin), argt
                           @nospecialize(rt))
     ⊑ = partialorder(𝕃)
     na = length(argtypes)
-    if f === memoryrefnew
+    if f === Core.memorynew
+        argtypes[1] isa Const && argtypes[2] isa Const || return false
+        MemT = argtypes[1].val
+        isconcretetype(MemT) && MemT <: GenericMemory || return false
+        len = argtypes[2].val
+        len isa Int && 0 <= len < typemax(Int) || return false
+        elsz = datatype_layoutsize(MemT)
+        checked_smul_int(len, elsz)[2] && return false
+        return true
```

Comment:
I find this kind of "compressed" code quite difficult to read and reason about. 
The rest of this function outlines the more complicated checks into their own functions?